File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Node{
        TreeNode node;
        int index;
        Node(TreeNode node , int index){
            this.node = node;
            this.index = index;
        }
    }
    public int widthOfBinaryTree(TreeNode root) {
        if(root == null)
            return 0;
      
        Deque<Node> queue = new ArrayDeque<>();
        queue.addLast(new Node(root,0));
        int result = 0;
        while(!queue.isEmpty()){
            int first = queue.peekFirst().index;
            int last = queue.peekLast().index;
            int l = queue.size() ;
            result = Math.max(result,last - first + 1 );
            for(int i=0;i < l;i++){
                Node ele = queue.pollFirst();
                if(ele.node.left != null)
                    queue.addLast(new Node(ele.node.left,ele.index * 2 +1));
                if(ele.node.right != null)
                    queue.addLast(new Node(ele.node.right,ele.index * 2 +2));
                

            }


        }
        return result;
        

        
    }
}

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0,[0] left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        queue = deque()
        queue.append([root,0])
        result = 0
        while queue:
            result = max(result , queue[-1][1] - queue[0][1] + 1)
            for i in range(len(queue)):
                node = queue.popleft()
                if node[0].left:
                    queue.append([node[0].left , node[1] * 2 + 1])
                if node[0].right:
                    queue.append([node[0].right , node[1] * 2 + 2])
                    
        return result



        